### PR TITLE
Cleanup and VMAD

### PR DESCRIPTION
**Pull request by pent0**
I can't test it yet since there is SECATTR bug that i don't know how to solve

After cleanup (moving and rewrite some functions)
image

### CODE REVIEW

File: src/emulator/shader/include/shader/usse_translator.h [hunk 1]
```diff
@@ -42,14 +43,117 @@ class USSETranslatorVisitor final {
...
+        // TODO: Array (OpAccessChain)
+        const SpirvVarRegBank &bank = get_reg_bank(op.bank);
+
+        // Composite a new vector
+        SpirvReg reg1;
+        std::uint32_t out_comp_offset;
+
+        bool result = bank.find_reg_at(op.num + offset, reg1, out_comp_offset);
+        if (!result) {
+            return 0xFFFFFFFF;
```
Comment (VelocityRa) [inline src/emulator/shader/include/shader/usse_translator.h]: `spv::NoResult`

File: src/emulator/shader/include/shader/usse_translator.h [hunk 2]
```diff
@@ -42,14 +43,117 @@ class USSETranslatorVisitor final {
...
+        }
+
+        if ((op.bank != USSE::RegisterBank::PRIMATTR && 
+            op.bank != USSE::RegisterBank::SECATTR && 
+            op.bank != USSE::RegisterBank::OUTPUT) || 
+            (out_comp_offset == 0)) {
```
Comment (VelocityRa) [inline src/emulator/shader/include/shader/usse_translator.h]: why parentheses?

File: src/emulator/shader/include/shader/usse_translator.h [hunk 3]
```diff
@@ -42,14 +43,117 @@ class USSETranslatorVisitor final {
...
+        , m_program_type(program_type) {
+        // Build common type here, so builder won't have to look it up later
+        spv_f32 = m_b.makeFloatType(32);
+        for (std::uint8_t i = 0 ; i < 4; i++) {
+            spv_vf32s[i] = m_b.makeVectorType(spv_f32, i + 1);
+        }
+    }
 
 private:
+    // Ry-o-ji ??? Is that you
```
Comment (VelocityRa) [inline src/emulator/shader/include/shader/usse_translator.h]: ಠ_ಠ

File: src/emulator/shader/include/shader/usse_translator.h [hunk 4]
```diff
@@ -42,14 +43,117 @@ class USSETranslatorVisitor final {
...
+    void store(Operand &dest, spv::Id source, std::uint8_t write_mask = 0xFF, std::uint8_t off = 0) {
+        const SpirvVarRegBank &bank = get_reg_bank(dest.bank);
+
+        // Composite a new vector
+        SpirvReg dest_reg;
+        std::uint32_t out_comp_offset;
+
+        bool result = bank.find_reg_at(dest.num + off, dest_reg, out_comp_offset);
+        if (!result) {
+            return;
```
Comment (VelocityRa) [inline src/emulator/shader/include/shader/usse_translator.h:136]: log something

File: src/emulator/shader/include/shader/usse_translator.h [hunk 5]
```diff
@@ -239,65 +324,76 @@ class USSETranslatorVisitor final {
...
-        for (auto repeat_offset = 0;
-             repeat_offset < repeat_count_num * repeat_offset_jump;
-             repeat_offset += repeat_offset_jump) {
-            SpirvReg src_reg;
-            SpirvReg dest_reg;
-            uint32_t src_comp_offset;
-            uint32_t dest_comp_offset;
+        // Is this VMAD3 or VMAD4, op2 = 0 => vec3
+        int type = 0;
+        (opcode2 == 0) ? (void)(type = 1) : (void)(type = 2);
```
Comment (VelocityRa) [inline src/emulator/shader/include/shader/usse_translator.h]: why the cast? just do a normal `if` imo

File: src/emulator/shader/include/shader/usse_translator.h [hunk 6]
```diff
@@ -239,65 +324,76 @@ class USSETranslatorVisitor final {
...
-    bool vpck(ExtPredicate pred, bool skipinv, bool nosched, Imm1 src2_bank_sel, bool syncstart, Imm1 end, Imm1 src1_ext_bank, Imm2 src2_ext_bank, RepeatCount repeat_count, Imm3 src_fmt, Imm3 dest_fmt, DestinationMask dest_mask, Imm2 src1_bank_sel, Imm5 dest_n, Imm2 comp_sel_3, Imm1 scale, Imm2 comp_sel_1, Imm2 comp_sel_2, Imm5 src1_n, Imm1 comp0_sel_bit1, Imm4 src2_n, Imm1 comp_sel_0_bit0) {
-        Instruction inst{};
+        // Write mask is a 4-bit immidiate
+        // If a bit is one, a swizzle is active
+        BEGIN_REPEAT(repeat_count, 2)
+            spv::Id vsrc0 = load(inst.opr.src0, repeat_offset, src0_abs, src0_neg);
+            spv::Id vsrc1 = load(inst.opr.src1, repeat_offset, gpi0_abs, gpi0_neg);
+            spv::Id vsrc2 = load(inst.opr.src2, repeat_offset, gpi1_abs, gpi1_neg);
+
+            if (vsrc0 == 0xFFFFFFFF || vsrc1 == 0xFFFFFFFF || vsrc2 == 0xFFFFFFFF) {
```
Comment (VelocityRa) [inline src/emulator/shader/include/shader/usse_translator.h]: check `spv::NoResult`

File: src/emulator/shader/include/shader/usse_translator.h [hunk 7]
```diff
@@ -239,65 +324,76 @@ class USSETranslatorVisitor final {
...
+                LOG_ERROR("Can't found reg!");
```
Comment (VelocityRa) [inline src/emulator/shader/include/shader/usse_translator.h]: `find` and print some more stuff ex. the reg it couldn't find

**Review by VelocityRa — CHANGES_REQUESTED** [review]
;o

File: src/emulator/shader/include/shader/usse_translator.h (hunk 3 above)
Comment (pent0) [reply src/emulator/shader/include/shader/usse_translator.h]: ...

File: src/emulator/shader/include/shader/usse_translator.h (hunk 5 above)
Comment (pent0) [reply src/emulator/shader/include/shader/usse_translator.h]: was lazy, will adjust it

**Pull request merged**